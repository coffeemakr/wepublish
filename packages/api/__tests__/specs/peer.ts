import {ApolloServerTestClient} from 'apollo-server-testing'
import {FetchMock} from 'jest-fetch-mock'
import fetch from 'node-fetch'
import {
  CreateNewsroom,
  CreateNewsroomInput,
  DeleteNewsroom,
  Newsroom,
  NewsroomList,
  PeerProfile,
  UpdateNewsroom,
  UpdateOwnNewsroom,
  UpdateOwnNewsroomInput
} from '../api/private'
import fakePeerAdminSchema from '../fakePeerAdminSchema.json'

import {createGraphQLTestClientWithPrisma, generateRandomString} from '../utility'
;((fetch as unknown) as FetchMock).mockResponse(JSON.stringify(fakePeerAdminSchema))

let testClientPrivate: ApolloServerTestClient

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithPrisma()
    testClientPrivate = setupClient.testClientPrivate
  } catch (error) {
    console.log('Error', error)

    throw new Error('Error during test setup')
  }
})

describe('Peers', () => {
  describe('can be created/edited/deleted:', () => {
    const ids: string[] = []
    beforeAll(async () => {
      const {mutate} = testClientPrivate
      const input: CreateNewsroomInput = {
        name: `Peer Test ${ids.length}`,
        slug: generateRandomString(),
        hostURL: 'https://host-url.ch/',
        token: `token${ids.length}`
      }
      const res = await mutate({
        mutation: CreateNewsroom,
        variables: {
          input: input
        }
      })
      ids.unshift(res.data.createPeer?.id)
    })

    test('can be created', async () => {
      const {mutate} = testClientPrivate
      const input: CreateNewsroomInput = {
        name: 'Create Peer Test',
        slug: generateRandomString(),
        hostURL: 'https://host-url.ch/',
        token: 'testTokenABC123'
      }
      const res = await mutate({
        mutation: CreateNewsroom,
        variables: {
          input: input
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          createNewsroom: {
            id: expect.any(String),
            slug: expect.any(String)
          }
        }
      })
      ids.unshift(res.data.createNewsroom?.id)
    })

    test('can be read in list', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: NewsroomList
      })

      expect(res.data.newsrooms).not.toHaveLength(0)
    })

    test('can be read by id', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: Newsroom,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          newsroom: {
            id: expect.any(String),
            slug: expect.any(String)
          }
        }
      })
    })

    test('can be updated', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: UpdateNewsroom,
        variables: {
          input: {
            name: 'Updated Peer',
            slug: generateRandomString(),
            hostURL: 'https://new-host-url.ch/',
            token: 'newTestTokenABC123'
          },
          id: ids[0]
        }
      })

      expect(res).toMatchSnapshot({
        data: {
          updateNewsroom: {
            id: expect.any(String),
            slug: expect.any(String)
          }
        }
      })
    })

    test('can be deleted', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: DeleteNewsroom,
        variables: {
          id: ids[0]
        }
      })

      expect(res).toMatchSnapshot({
        data: {
          deleteNewsroom: expect.any(Object)
        }
      })
      expect(res.data.deleteNewsroom.id).toBe(ids[0])

      ids.shift()
    })

    test('can read peer profile', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: PeerProfile
      })
      expect(res).toMatchSnapshot({
        data: {
          peerProfile: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can update peer profile', async () => {
      const {mutate} = testClientPrivate
      const input: UpdateOwnNewsroomInput = {
        name: 'test peer profile',
        logoID: 'logoID123',
        themeColor: '#4287f5',
        themeFontColor: '#d67c15',
        callToActionText: [{text: 'rich text call to action'}],
        callToActionURL: 'calltoactionurl.ch/'
      }

      const res = await mutate({
        mutation: UpdateOwnNewsroom,
        variables: {
          input: input
        }
      })

      expect(res).toMatchSnapshot()
    })
  })
})
