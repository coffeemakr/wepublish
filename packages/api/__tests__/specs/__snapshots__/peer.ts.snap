// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Peers can be created/edited/deleted: can be created 1`] = `
Object {
  "data": Object {
    "createNewsroom": Object {
      "callToActionImage": null,
      "callToActionImageURL": null,
      "callToActionText": Array [],
      "callToActionURL": null,
      "hostURL": "https://host-url.ch/",
      "id": Any<String>,
      "isDisabled": false,
      "isSelf": false,
      "logo": null,
      "name": "Create Peer Test",
      "slug": Any<String>,
      "themeColor": "#000000",
      "themeFontColor": "#FFFFFF",
      "token": "testTokenABC123",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be deleted 1`] = `
Object {
  "data": Object {
    "deleteNewsroom": Any<Object>,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be read by id 1`] = `
Object {
  "data": Object {
    "newsroom": Object {
      "callToActionImage": null,
      "callToActionImageURL": null,
      "callToActionText": Array [],
      "callToActionURL": null,
      "hostURL": "https://host-url.ch/",
      "id": Any<String>,
      "isDisabled": false,
      "isSelf": false,
      "logo": null,
      "name": "Create Peer Test",
      "slug": Any<String>,
      "themeColor": "#000000",
      "themeFontColor": "#FFFFFF",
      "token": "testTokenABC123",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can be updated 1`] = `
Object {
  "data": Object {
    "updateNewsroom": Object {
      "callToActionImage": null,
      "callToActionImageURL": null,
      "callToActionText": Array [],
      "callToActionURL": null,
      "hostURL": "https://new-host-url.ch/",
      "id": Any<String>,
      "isDisabled": false,
      "isSelf": false,
      "logo": null,
      "name": "Updated Peer",
      "slug": Any<String>,
      "themeColor": "#000000",
      "themeFontColor": "#FFFFFF",
      "token": "newTestTokenABC123",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can read peer profile 1`] = `
Object {
  "data": Object {
    "peerProfile": Object {
      "callToActionImage": null,
      "callToActionImageURL": null,
      "callToActionText": Array [],
      "callToActionURL": null,
      "id": Any<String>,
      "isSelf": true,
      "logo": null,
      "name": "Add your newsroom name here",
      "themeColor": "#000000",
      "themeFontColor": "#FFFFFF",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Peers can be created/edited/deleted: can update peer profile 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: 
Invalid \`prisma.newsroom.update()\` invocation:


  Foreign key constraint failed on the field: \`newsrooms_logoID_fkey (index)\`],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
